flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java

```diff
@@ -825,6 +828,78 @@ public void testFatalErrorAfterUnInterruptibleInvoke() throws Exception {
 		}
 	}
 
+	/**
+	 * Tests that a fatal error gotten from canceling task is notified.
+	 */
+	@Test
+	public void testFatalErrorOnCanceling() throws Exception {
+		final AwaitFatalErrorTaskManagerActions taskManagerActions =
+			new AwaitFatalErrorTaskManagerActions();
+
+		final Configuration config = new Configuration();
+		config.setLong(TaskManagerOptions.TASK_CANCELLATION_INTERVAL, 5);
+		config.setLong(TaskManagerOptions.TASK_CANCELLATION_TIMEOUT, 50);
+
+		final Task task = spy(createTaskBuilder()
+			.setInvokable(InvokableBlockingWithTrigger.class)
+			.setTaskManagerConfig(config)
+			.setTaskManagerActions(taskManagerActions)
+			.build());
+
+		doThrow(OutOfMemoryError.class).when(task).cancelOrFailAndCancelInvokableInternal(eq(ExecutionState.CANCELING), eq(null));
+
+		try {
+			task.startTaskThread();
+
+			awaitLatch.await();
+
+			task.cancelExecution();
+
+			// wait for the notification of notifyFatalError
+			taskManagerActions.latch.await();
+		} catch (Throwable t) {
+			fail("No exception is expected to be thrown by fatal error handling");
+		} finally {
+			triggerLatch.trigger();
+		}
+	}
+
+	/**
+	 * Tests that a fatal error gotten from externally failing task is notified.
+	 */
+	@Test
+	public void testFatalErrorOnFailingExternally() throws Exception {
```

Comment:
Do we need both of those tests?

If so can we deduplicate the code of those two tests into a common method that takes to be thrown exception class and/or target state as parameters?

Reply:
I added the second test just to increase the coverage, but I do not think the second one is strictly necessary. In general, I try to follow the style of the same test class module (TaskTest), which has little factored code and I think is better for test codes. I can factor out the common ones, but I would just better remove the second test, as long as you agree.

Reply:
> I added the second test just to increase the coverage

~how was it increasing the test coverage? What was the difference between them?~

edit: oh, I see it now. One was calling `cancelExecution` and another `failExternally`. Ehhh, since this test is already relaying very heavily on the details of the implementation, I would say it's fine to remove one of them.